barretenberg/cpp/src/barretenberg/client_ivc/client_ivc.cpp

```diff
@@ -213,40 +213,103 @@ void ClientIVC::accumulate(ClientCircuit& circuit, const std::shared_ptr<Verific
 }
 
 /**
- * @brief Construct a proof for the IVC, which, if verified, fully establishes its correctness
+ * @brief Construct the hiding circuit, which recursively verifies the last folding proof and decider proof, and
+ * then produce a proof of the circuit's correctness with MegaHonk.
  *
- * @return Proof
+ * @details The aim of this intermediate stage is to reduce the cost of producing a zero-knowledge ClientIVCProof.
+ * @return HonkProof - a Mega proof
  */
-ClientIVC::Proof ClientIVC::prove()
+HonkProof ClientIVC::construct_and_prove_hiding_circuit()
 {
     max_block_size_tracker.print();               // print minimum structured sizes for each block
     ASSERT(verification_queue.size() == 1);       // ensure only a single fold proof remains in the queue
     ASSERT(merge_verification_queue.size() == 1); // ensure only a single merge proof remains in the queue
+
     FoldProof& fold_proof = verification_queue[0].proof;
-    MergeProof& merge_proof = merge_verification_queue[0];
     HonkProof decider_proof = decider_prove();
-    // Free the accumulator to save memory
+
     fold_output.accumulator = nullptr;
-    return { fold_proof, std::move(decider_proof), goblin.prove(merge_proof) };
+
+    ClientCircuit builder{ goblin.op_queue };
+    // The last circuit being folded is a kernel circuit whose public inputs need to be passed to the base rollup
+    // circuit. So, these have to be preserved as public inputs to the hiding circuit (and, subsequently, as public
+    // inputs to the tube circuit) which are intermediate stages.
+    // TODO(https://github.com/AztecProtocol/barretenberg/issues/1048): link these properly, likely insecure
+    auto num_public_inputs = static_cast<uint32_t>(static_cast<uint256_t>(fold_proof[1]));
+    vinfo("num_public_inputs of the last folding proof BEFORE SUBTRACTION", num_public_inputs);
+    num_public_inputs -= bb::AGGREGATION_OBJECT_SIZE;             // exclude aggregation object
+    num_public_inputs -= bb::PROPAGATED_DATABUS_COMMITMENTS_SIZE; // exclude propagated databus commitments
+    vinfo("num_public_inputs of the last folding proof ", num_public_inputs);
+    for (size_t i = 0; i < num_public_inputs; i++) {
+        size_t offset = 3;
```

Comment:
Is it possible to use the constant `HONK_RECURSION_PUBLIC_INPUT_OFFSET` instead of the magic 3 here? I know you didnt introduce this